File: parallelism_test.py
import threading
import time

class WorkerThread(threading.Thread):
    def __init__(self, id):
        super().__init__()
        self.id = id
        self.active = True
        self.pause_condition = threading.Condition(threading.Lock())  # Use a condition to manage pausing
        self.paused = False

    def run(self):
        while self.active:
            with self.pause_condition:
                while self.paused:
                    self.pause_condition.wait()  # Release lock and wait for notification

            print(f"ID: {self.id}, Time: {time.ctime()}")

            # Busy-wait for about 1 second
            end_time = time.time() + 1
            while time.time() < end_time:
                pass

    def stop(self):
        """ Stop the thread completely """
        self.active = False
        # If paused, resume the thread to ensure it can be stopped
        if self.paused:
            self.resume()

    def pause(self):
        """ Pause the thread """
        with self.pause_condition:
            self.paused = True

    def resume(self):
        """ Resume the thread """
        with self.pause_condition:
            self.paused = False
            self.pause_condition.notify()  # Unpause the paused thread


def main():
    # Start 5 worker threads
    threads = [WorkerThread(i) for i in range(5)]
    for t in threads:
        t.start()

    time.sleep(3)  # Let the threads run for 5 seconds

    # Pause the first 2 threads
    print("\nPausing first 2 threads\n")
    for i in range(2):
        threads[i].pause()

    time.sleep(3)  # Let the remaining threads run for 5 more seconds

    # Restart the paused threads
    print("\nRestarting the paused threads\n")
    for i in range(2):
        threads[i].resume()

    # Let them run for another 5 seconds
    time.sleep(3)

    # Stop all threads
    for t in threads:
        t.stop()
        t.join()

    print("All threads have been stopped")


if __name__ == '__main__':
    main()

File: parallelism_test.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <pthread.h>
#include <time.h>

#define NUM_THREADS 5

typedef struct {
    int id;
    int active;
    pthread_mutex_t pause_mutex;
    pthread_cond_t pause_cond;
    int paused;
} WorkerData;

void* worker_thread(void* arg) {
    WorkerData* data = (WorkerData*) arg;
    clock_t start_time, end_time;

    while(data->active) {
        pthread_mutex_lock(&data->pause_mutex);
        while(data->paused) {
            pthread_cond_wait(&data->pause_cond, &data->pause_mutex);
        }
        pthread_mutex_unlock(&data->pause_mutex);

        time_t now;
        time(&now);
        printf("ID: %d, Time: %s", data->id, ctime(&now));

        // Busy-wait loop for about 1 second
        start_time = clock();
        do {
            end_time = clock();
        } while((double)(end_time - start_time) / CLOCKS_PER_SEC < 1.0);
    }

    return NULL;
}

void pause_worker(WorkerData* data) {
    pthread_mutex_lock(&data->pause_mutex);
    data->paused = 1;
    pthread_mutex_unlock(&data->pause_mutex);
}

void resume_worker(WorkerData* data) {
    pthread_mutex_lock(&data->pause_mutex);
    data->paused = 0;
    pthread_cond_signal(&data->pause_cond);
    pthread_mutex_unlock(&data->pause_mutex);
}

int main() {
    pthread_t threads[NUM_THREADS];
    WorkerData thread_data[NUM_THREADS];

    // Initialize and start threads
    for(int i = 0; i < NUM_THREADS; i++) {
        thread_data[i].id = i;
        thread_data[i].active = 1;
        thread_data[i].paused = 0;
        pthread_mutex_init(&thread_data[i].pause_mutex, NULL);
        pthread_cond_init(&thread_data[i].pause_cond, NULL);
        pthread_create(&threads[i], NULL, worker_thread, &thread_data[i]);
    }

    sleep(5);

    // Pause the first 2 threads
    printf("\nPausing first 2 threads\n");
    for(int i = 0; i < 2; i++) {
        pause_worker(&thread_data[i]);
    }

    sleep(5);

    // Resume the first 2 threads
    printf("\nResuming paused threads\n");
    for(int i = 0; i < 2; i++) {
        resume_worker(&thread_data[i]);
    }

    sleep(5);

    // Stop all threads
    for(int i = 0; i < NUM_THREADS; i++) {
        thread_data[i].active = 0;
    }

    for(int i = 0; i < NUM_THREADS; i++) {
        pthread_join(threads[i], NULL);
        pthread_mutex_destroy(&thread_data[i].pause_mutex);
        pthread_cond_destroy(&thread_data[i].pause_cond);
    }

    printf("All threads have been stopped\n");
    return 0;
}
